spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,186 @@
+4. The Application checks the block (the header and transactions). It can also:
+   * add/remove transactions
+   * modify the header hashes
+   * if the block is modified, the Application sets it in the return parameters
+5. The Application signals _Accept_ of _Reject_ in `PrepareProposal`'s return values
+   * TODO: Decide if this kind if Accept/Reject is wanted/covered by impl (maybe a panic here makes more sense?)
+   * If _Reject_, the proposed block --along with any modification-- is discarded. Tendermint interprets that there is nothing to be proposed for consensus at the moment and _p_ won't send any proposal message in round _r_, height _h_.
+   * If _Accept_, _p_'s Tendermint uses the modified block as _p_'s proposal in round _r_, height _h_.
+
+### ProcessProposal
+
+#### Parameters and Return values
+
+TODO
+
+#### When does Tendermint call it?
+
+When a validator _p_ enters Tendermint consensus round _r_, height _h_, in which _q_ is the proposer (possibly _p_ = _q_):
+
+1. _p_ sets up timer `ProposeTimeout`.
+2. If _p_ is the proposer, execute steps 1-6 in _PrepareProposal_.
+3. Upon reception of `Proposal` message from _q_ for round _r_, height _h_, _p_'s Tendermint calls `ProcessProposal` with the newly received proposal. The call is synchronous.
+4. The Application checks/processes the proposed block, which is read-only, and returns _Accept_ or _Reject_.
+   * Depending on the Application's needs, it may return from `ProcessProposal`
+     * either after it has completely processed the block (the simpler case),
+     * or immediately (after doing some basic checks), and process the block asynchronously. In this case the Application won't be able to reject the block, or force prevote/precommit `nil` afterwards.
+5. If the returned value is
+     * _Reject_, Tendermint will prevote `nil` for the proposal in round _r_, height _h_ (this can be interpreted as _valid(v)_ returning false in the white paper's pseudocode)
+     * _Accept_, Tendermint will follow the normal algorithm to prevote on the proposal in _r_, height _h_
+
+### ExtendVote
+
+#### Parameters and Return values
+
+TODO
+
+#### When does Tendermint call it?
+
+When a validator _p_ is in Tendermint consensus round _r_, height _h_, state _prevote_, and the other conditions prescribed by the consensus algorithm for sending a _Precommit_ message are fulfilled:
+
+1. _p_'s Terndermint updates the consensus state as prescriped by the algorithm (_validValue_, _lockedValue_, etc.)
+2. _p_'s Tendermint calls `ExtendVote` with the value that is about to be sent as precommit message. The call is synchronous.
+3. The Application returns an array of bytes, `precommit_extension`, which is not interpreted by Tendermint.
+4. _p_'s Tendermint includes `precommit_extension` as a new field in the `precommit` message.
+5. _p_'s Tendermint broadcasts the `precommit` message.
```

Comment:
Can it happen that a validator is byzantine and equivocates? Perhaps only on the `precommit_extension`?

Reply:
Indeed. That is something we need to explore IMO. But first, I think I need a discussion with Dev (and others?) to understand how exactly they intend to use this field.

Reply:
Discussion: we came up with a set of rules for ExtendVote, which will be included in the next commit to this PR